### ISSUE TITLE
Use unsigned comparators for TR::Address in jitBuilder

### ISSUE THREAD

**Issue by liqunl** [eclipse/omr#1225]
jitBuilder is calling to `OMR::ILOpCode::compareOpCode` to ask for signed comparators for `TR::Address` type, which should be using unsigned compartors. Code in jitBuilder should be asking for unsigned comparators for `TR::Address` and switch cases for `TR::Address` should be removed from the signed group in `OMR::ILOpCode::compareOpCode`. More thoughts should be given to (in)equality compares since they are sign insensitive.

**Comment by mstoodle** [comment]
The signed comparators (LessThan, etc.) ask for signed opcodes. The unsigned comparators (UnsignedLessThan) ask for unsigned opcodes. Because the comparison operations are returned from TR::ILOps::compareOpCode under the "signed" branch rather than the "unsigned" branch, JitBuilder clients are being encouraged to compare pointers using signed comparison services.

If the underlying opcodes are only provided for unsigned comparisons and assert for signed comparisons, then JitBuilder clients will be encouraged to do unsigned comparisons on pointers.

The (in)equality point is mildly interesting, but probably mostly pedantic. Given (in)equality is sign insensitive, why would compareOpCode() care about the value of the 'unsignedCompare' parameter? Shouldn't the acmp{eq,ne} opcodes be returned for TR::Address regardless of whether unsignedCompare is true or false?

In short, I'm not sure there's anything for JitBuilder *itself* to do here, though some clients could be briefly annoyed by the change from wrong to correct behaviour requiring them to change their code.

**Comment by liqunl** [comment]
I'm thinking maybe comparators like `(if)acmp*` should not care about `unsignedCompare` parameter since they only have one version. Other examples are comparators for `TR::Float` and `TR::Double`. There is only one signedness for those types, so are their comparators. And clients should know that.

**Comment by mstoodle** [comment]
But does the client always know that it's asking for a compare opcode for Address/Float/Double type? The JitBuilder service calling compareOpCode does not, so that means every client of compareOpCode would have to write a bunch of conditional code to properly figure out whether signed/unsigned matters.

I would prefer compareOpCode (and its brethren for other kinds of opcodes) to just return the right thing, so that if the notion of "the right thing" evolves or if the set of data types expands, compareOpCode only needs to evolve in one place.

**Comment by liqunl** [comment]
I think the client should know the type of data they are comparing. For JitBuilder, based on the fact that it has successfully avoided crashes on Address/Float/Double compares, I would assume it knows the type of data it compares. The client might not know we don't respect `unsignedCompare` and thus can make wrong assumptions. 

Maybe a more proper solution could be removing `unsignedCompare` from `compareOpCode` argument list and have `compareOpCode` return the default comparators. Then whoever need a unsigned comparator should convert the comparator to an unsigned one themselves. And they should not ask for an unsigned comparator on Float/Double because it's not supported, not in any language that I know. Same as Address who does not have signed version. This way, we don't need to spell out nor worry about the signedness of (in)equality compares.

**Comment by mstoodle** [comment]
I would prefer we don't create a solution that sprays a somewhat subtle conditional equation to every code location that wants to build a compare node. That sounds to me like a recipe for bugs to creep in.

So I think I would prefer we keep compareOpCode and avoid making the question more complicated to ask. We can reward the nonsensical questions with assertions.

I think this work item resides solely within compareOpCode. The only connection to JitBuilder is that any client who asks a currently nonsensical question may start throwing assertions, but that can be fixed by asking the right question. Possibly the JitBuilder services could throw the assertion as an improvement, but I don't usually like implementing duplicate assertions, and I think that's a larger question for JitBuilder than what happens with comparisons on TR::Address'es.

**Comment by liqunl** [comment]
I double check the code, not many code need to be modified. The only place I've seen is asking for (in)equality comparators on `TR::Address`. For (in)equality, we can always return the right thing. Even if there are two versions of comparators for integers, the meanings of them are same. 

I agree we should reward the nonsensical questions with assertions so that the client would know what's wrong right away.

Changes might not be needed for JitBuilder if it's only using (in)equality comparators on address. I'll do a test to confirm that.

**Comment by lmaisons** [comment]
Back when we were working on Python IL generation, I had a conversation with Pat [Doyle] about ordering comparisons between pointers, as the evaluators for such operations hadn't been implemented [at least on x86_64].  If I remember correctly, he was of the opinion that pointer comparison could / should be considered a signed operation [again, at least on x86_64] and recommended that I attempt to use the signed long evaluator.

On the other hand, on x86_32 3GB and 4GB user areas seem to pretty much require that pointers on that platform be considered unsigned...

I believe that on any given platform there is only one version of 'the right answer' when it comes to address comparisons, thus expecting the user to distinguish between signed and unsigned is a recipe for madness.  To the point where I wonder if the best answer would be to provide two overloads to `OMR::ILOpCode::compareOpCode` one that takes a sign parameter and one that doesn't, and assert in the former when the distinction is nonsensical, and in the latter when it's critical.
